corehq/apps/reports/standard/tableau.py

```diff
@@ -63,34 +64,47 @@ def page_context(self):
         if self.visualization.server.server_type == 'server':
-            return render(self.request, self.template_name)
+            return render(self.request, self.template_name, self.get_context_data())
         return super().get(request, *args, **kwargs)
 
-    def tableau_server_response(self):
-        from requests_toolbelt.adapters import host_header_ssl  # avoid top-level import that breaks docs build
-        context = self.get_context_data()
-        tabserver_url = 'https://{}/trusted/'.format(self.visualization.server.server_name)
-        post_arguments = {'username': self.visualization.server.domain_username}
-        if self.visualization.server.target_site != 'Default':
-            post_arguments.update({'target_site': self.visualization.server.target_site})
-
-        if self.visualization.server.validate_hostname != '':
-            request = requests.Session()
-            request.mount('https://', host_header_ssl.HostHeaderSSLAdapter())
-            tabserver_response = request.post(tabserver_url,
-                                              post_arguments,
-                                              headers={'Host': self.visualization.server.validate_hostname})
-        else:
-            tabserver_response = requests.post(tabserver_url, post_arguments)
-
-        if tabserver_response.status_code == 200:
-            if tabserver_response.content != b'-1':
-                context.update({'ticket': tabserver_response.content.decode('utf-8')})
-                return render(self.request, self.template_name, context)
-            else:
-                context.update({"failure_message": _("Tableau trusted authentication failed")})
-                return render(self.request, 'reports/tableau_server_request_failed.html', context)
+
+@require_POST
+def tableau_visualisation_ajax(request, domain):
+    from requests_toolbelt.adapters import host_header_ssl
+    visualiation_data = {data: value[0] for data, value in dict(request.POST).items()}
+    server_name = visualiation_data.pop('server_name')
+    validate_hostname = visualiation_data.pop('validate_hostname')
+    target_site = visualiation_data.pop('target_site')
+    domain_username = visualiation_data.pop('domain_username')
+
+    tabserver_url = 'https://{}/trusted/'.format(server_name)
+    post_arguments = {'username': domain_username}
+
+    if target_site != 'Default':
+        post_arguments.update({'target_site': target_site})
+
+    if validate_hostname != '':
+        request = requests.Session()
+        request.mount('https://', host_header_ssl.HostHeaderSSLAdapter())
+        tabserver_response = request.post(tabserver_url,
+                                          post_arguments,
+                                          headers={'Host': validate_hostname})
+    else:
+        tabserver_response = requests.post(tabserver_url, post_arguments)
+
+    if tabserver_response.status_code == 200:
+        if tabserver_response.content != b'-1':
```

Comment:
lol, that's annoying - it gives a 200 response when auth fails?

Reply:
haha yes it just sends an invalid ticket... 🙈